engine/runtime/src/main/java/org/enso/interpreter/runtime/data/EnsoDate.java

```diff
@@ -0,0 +1,128 @@
+package org.enso.interpreter.runtime.data;
+
+import com.oracle.truffle.api.CompilerDirectives;
+import com.oracle.truffle.api.dsl.Cached;
+import com.oracle.truffle.api.dsl.Specialization;
+import com.oracle.truffle.api.interop.InteropLibrary;
+import com.oracle.truffle.api.interop.TruffleObject;
+import com.oracle.truffle.api.library.CachedLibrary;
+import com.oracle.truffle.api.library.ExportLibrary;
+import com.oracle.truffle.api.library.ExportMessage;
+
+import java.time.DateTimeException;
+import java.time.LocalDate;
+import java.time.format.DateTimeFormatter;
+import java.time.format.DateTimeParseException;
+
+import org.enso.interpreter.dsl.Builtin;
+import org.enso.interpreter.node.expression.builtin.error.PolyglotError;
+import org.enso.interpreter.runtime.Context;
+import org.enso.interpreter.runtime.callable.UnresolvedSymbol;
+import org.enso.interpreter.runtime.callable.function.Function;
+import org.enso.interpreter.runtime.data.text.Text;
+import org.enso.interpreter.runtime.library.dispatch.MethodDispatchLibrary;
+
+@ExportLibrary(InteropLibrary.class)
+@ExportLibrary(MethodDispatchLibrary.class)
+@Builtin(pkg = "date", name = "Date")
+public final class EnsoDate implements TruffleObject {
+  private final LocalDate date;
+
+  public EnsoDate(LocalDate date) {
+    this.date = date;
+  }
+
+  @Builtin.Method(description = "Return current Date")
+  public static EnsoDate now() {
+    return new EnsoDate(LocalDate.now());
+  }
+
+  @Builtin.Method(name = "internal_parse", description = "Constructs a new Date from text with optional pattern")
+  @Builtin.Specialize
+  @Builtin.WrapException(from = DateTimeParseException.class, to = PolyglotError.class, propagate = true)
+  public static EnsoDate parse(Text text, Object noneOrPattern) {
+    var str = text.getContents().toString();
+    if (noneOrPattern instanceof Text pattern) {
+      noneOrPattern = pattern.getContents().toString();
+    }
```

Comment:
Will it ever be possible that `noneOrPattern` is a `String` instead of `Text`?

Reply:
Good question! Let's try it: 67e9256abfe786c5f1063ed477a29ea3593e4cf3